### PR TITLE
Mech & Tag > Property Part 1

### PR DESCRIPTION
**Pull request by MC-Samuel**
Part 1 of some Entity Mech and Tag merges.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java [hunk 1]
```diff
@@ -5,89 +5,67 @@
...
-            }
-            else if (entity.isVindicator()) {
-                return new ElementTag(entity.getVindicator().isJohnny());
-            }
-            return null;
-        });
+        autoRegister("angry", EntityAngry.class, ElementTag.class, false);
     }
 
+
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java]: stray edit

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java [hunk 2]
```diff
@@ -5,89 +5,67 @@
...
     }
 
     EntityTag entity;
 
     @Override
-    public String getPropertyString() {
+    public ElementTag getPropertyValue() {
         if (isWolf()) {
-            return getWolf().isAngry() ? "true" : null;
+            return new ElementTag(getWolf().isAngry());
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java]: this is a behavioral change

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 3]
```diff
@@ -3,115 +3,78 @@
...
-    public static final String[] handledMechs = new String[] {
-            "anger"
-    };
+    EntityTag entity;
 
-    public EntityAnger(EntityTag entity) {
-        this.entity = entity;
+    @Override
+    public DurationTag getPropertyValue() {
+        if (entity.getBukkitEntity() instanceof Bee) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: whyyy did you do that

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 4]
```diff
@@ -3,115 +3,78 @@
...
+    EntityTag entity;
 
-    public EntityAnger(EntityTag entity) {
-        this.entity = entity;
+    @Override
+    public DurationTag getPropertyValue() {
+        if (isBee()) {
+            return new DurationTag((long) getBee().getAnger());
+        }
+        else if (isPigZombie()) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: why are these checks different from before?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java [hunk 5]
```diff
@@ -5,87 +5,64 @@
...
     }
 
     EntityTag entity;
 
     @Override
-    public String getPropertyString() {
+    public ElementTag getPropertyValue() {
         if (isWolf()) {
-            return getWolf().isAngry() ? "true" : null;
+            return new ElementTag(getWolf().isAngry() ? "true" : "null");
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java]: wtf no

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 6]
```diff
@@ -3,115 +3,81 @@
...
+    public void setPropertyValue(DurationTag param, Mechanism mechanism) {
+        if (mechanism.requireObject(DurationTag.class)) {
+            DurationTag duration;
+            if (mechanism.getValue().isInt()) { // Soft-deprecated - backwards compatibility, as this used to use a tick count
+                duration = new DurationTag(mechanism.getValue().asLong());
+            }
+            else {
+                duration = mechanism.valueAsType(DurationTag.class);
+            }
+            if (isBee()) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: why is this changed from what it was before

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java [hunk 7]
```diff
@@ -5,87 +5,64 @@
...
-        // @returns ElementTag(Boolean)
-        // @mechanism EntityTag.angry
-        // @group properties
-        // @description
-        // If the entity is a wolf or PigZombie, returns whether the entity is angry.
-        // If the entity is a Vindicator, returns whether it is in "Johnny" mode.
-        // -->
-        PropertyParser.registerTag(EntityAngry.class, ElementTag.class, "angry", (attribute, entity) -> {
-            if (entity.isWolf()) {
-                return new ElementTag(entity.getWolf().isAngry());
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAngry.java]: did you test that the tag still works exactly the same as before (spoiler: it doesn't, you changed it)

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 8]
```diff
@@ -3,115 +3,81 @@
...
+        return new DurationTag((long) getAnger());
     }
 
-    EntityTag entity;
-
     @Override
-    public String getPropertyString() {
-        return new DurationTag((long) getAnger()).identify();
+    public void setPropertyValue(DurationTag param, Mechanism mechanism) {
+        if (mechanism.requireObject(DurationTag.class)) {
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: Same here, no need for `requireObject`/`valueAsType`, since `autoRegister` does all of that internally & you already get a valid and checked `DurationTag` here

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityArms.java [hunk 9]
```diff
@@ -3,80 +3,40 @@
...
-    EntityTag dentity;
-
     @Override
-    public String getPropertyString() {
-        return getStand().hasArms() ? "true" : null;
+    public void setPropertyValue(ElementTag param, Mechanism mechanism) {
+        if (mechanism.requireBoolean()) {
+            as(ArmorStand.class).setArms(param.asBoolean());
+        }
     }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityArms.java:37]: The old `getPropertyString` returned `null` for false, as that was the default value (I.e. didn't need to be specified in the entity's property map) - `getPropertyValue` should still return the full boolean so that the `.arms` tag works as expected, but should implement `isDefaultValue` here.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityBasePlate.java [hunk 10]
```diff
@@ -3,44 +3,32 @@
...
     }
 
     @Override
-    public String getPropertyString() {
-        return getStand().hasBasePlate() ? null : "false";
+    public void setPropertyValue(ElementTag param, Mechanism mechanism) {
+        if (mechanism.requireBoolean()) {
+            as(ArmorStand.class).setBasePlate(param.asBoolean());
+        }
     }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityBasePlate.java:37]: Same here but the other way around, `true` is the default value

**Review by tal5 — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 11]
```diff
@@ -3,115 +3,81 @@
...
-        }
+    public static boolean describes(EntityTag entity) {
+        return entity.getBukkitEntity() instanceof Bee
+                || entity.getBukkitEntity() instanceof PigZombie;
     }
 
-    public static final String[] handledMechs = new String[] {
-            "anger"
-    };
+    EntityTag entity;
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: My bad if I'm missing something, but was this tested? it looks like the old field was kept but isn't ever set, I.e. the entity is always `null`.
Generally the old fields, `is/getPigZombie` sorta methods, and stuff like `getAnger` can just be removed in favor of doing it directly with stuff like `instanceof` pattern matching, like:
```java
if (entity instanceof ZombiePig zombie) {
    zombie.doThing()
}
else {
    as(Bee).doThing();
}
```
But should at least remove the old field and make the existing methods use `EntityProperty#getEntity`.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 12]
```diff
@@ -3,115 +3,51 @@
...
-            return getPigZombie().getAnger();
+    public void setPropertyValue(DurationTag param, Mechanism mechanism) {
+        if (getEntity() instanceof PigZombie) {
+            as(PigZombie.class).setAnger(param.getTicksAsInt());
         }
         else {
-            return getBee().getAnger();
+            as(Bee.class).setAnger(param.getTicksAsInt());
         }
     }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: Missing the backwards compatibility for old input the mechanism had (can probably just have an if check to reassign `param`)

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 13]
```diff
@@ -3,115 +3,51 @@
...
-    public String getPropertyId() {
-        return "anger";
-    }
-
-    public int getAnger() {
-        if (isPigZombie()) {
-            return getPigZombie().getAnger();
+    public void setPropertyValue(DurationTag param, Mechanism mechanism) {
+        if (getEntity() instanceof PigZombie) {
+            as(PigZombie.class).setAnger(param.getTicksAsInt());
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: Should use `instanceof` pattern matching

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 14]
```diff
@@ -3,115 +3,51 @@
...
     }
 
-    EntityTag entity;
-
     @Override
-    public String getPropertyString() {
-        return new DurationTag((long) getAnger()).identify();
+    public DurationTag getPropertyValue() {
+        if (getEntity() instanceof PigZombie pigzombie) {
+            return new DurationTag((long) pigzombie.getAnger());
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: Java naming conventions - should be `pigZombie`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 15]
```diff
@@ -3,115 +3,51 @@
...
+        }
+        else if (getEntity() instanceof Bee bee) {
+            return new DurationTag((long) bee.getAnger());
+        }
+        return null;
     }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java]: Just a nitpick, but don't need to have an `instanceof` check for bees here, as the entity is necessarily either a bee or a zombie piglin (I.e. can just `as(Bee.class)` like you did in the mechanism)

**Review by tal5 — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java [hunk 16]
```diff
@@ -3,115 +3,56 @@
...
-    public int getAnger() {
-        if (isPigZombie()) {
-            return getPigZombie().getAnger();
+    public void setPropertyValue(DurationTag param, Mechanism mechanism) {
+        if (mechanism.getValue().isInt()) { // Soft-deprecated - backwards compatibility, as this used to use a tick count
+            param = new DurationTag(mechanism.getValue().asLong());
         }
         else {
-            return getBee().getAnger();
+            param = mechanism.valueAsType(DurationTag.class);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAnger.java:40]: I believe this `else{}` block setter is redundant in the new `setPropertyValue` context

Please also test thoroughly to validate that the legacy tick input works as intended, since the code is used in a different context here

**Comment by mcmonkey4eva** [comment]
... You did not think through nor test that commit.
